### PR TITLE
Use pexpect to manage the MATLAB console

### PR DESCRIPTION
**Pull request by tdy**
Now calls `pexpect.spawn.interact()` for full command window functionality (properly handles arrow keys, tab completion, backspace, `clc`, etc.). Falls back to the original `sys.stdin` code if `pexpect` isn't available.

Note a couple side effects before deciding whether to merge:
* `print` causes flushing issues with `pexpect.spawn.interact()`, so now there's `print_flush()` as a workaround.
* Timer code is now echoed to the console (example attached). I couldn't figure out how to hide these when using `pexpect.spawn.send()`.

Example

### CODE REVIEW

**Comment by tdy** [comment]
Same as #12. Not sure what happened there. I must've messed up some commits accidentally.

File: scripts/vim-matlab-server.py [hunk 1]
```diff
@@ -76,36 +91,46 @@ def handle(self):
...
+
+def start_thread(target=None, args=()):
+    thread = threading.Thread(target=target, args=args)
+    thread.daemon = True
+    thread.start()
+
+def print_flush(value):
+    """Manually flush the line if using pexpect."""
+    if use_pexpect:
+        value += ''.join('\b' for _ in range(len(value)))
```
Comment (daeyun) [inline scripts/vim-matlab-server.py]: nit: Is this the same as

```python
value += '\b' * len(value)
```

**Comment by daeyun** [comment]
Thanks, this is definitely a very nice feature to add. The raw code being echoed seems to be an inherent issue coming from the fact that this runs a repl inside a pseudo-terminal. My guess is that `send` simulates actual keyboard input, and that causes the matlab console to print them out as if those keys were being pressed (though I don't really know about how pseudo-terminals work and haven't used pexpect before).

**Comment by daeyun** [comment]
I will accept the pull request. I think there might be a way to use `pexpect.spawn.interact()`'s `output_filter` to filter out any strings that match the ones sent to `pexpect.spawn.send()`. I guess that can be addressed in a different patch.

**Comment by daeyun** [comment]
0f8a799 was my attempt. Hopefully this works, but there might be edge cases that I haven't checked.

I think an idiomatic way to do this sort of thing might be to do something like `pexpect.spawn.expect_exact(command)` after `pexpect.spawn.send(command)` so that it consumes the next occurrence of "command". But for some reason, it didn't seem to work with the MATLAB repl. So I ended up using an `output_filter` that ignores outputs until it sees a newline character.

**Comment by tdy** [comment]
Ah nice, it works for the most part.

Something weird happens with certain code though. I haven't had a chance to debug the cause, but one example is this dummy cell:

```matlab
%%
magic(5)

2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
2987.293487298 + 84.7928748123 + 234 + 234.293847293 + 39847 - 29384798 - 234.1239871;
```

It not only prints the raw code, but also truncates it:
```
>> NMITNNQHQSXI=tic;magic(5),2987.293487298 + 84.7928748123 + 234 + 234.293847293 +
39847 - 29384798 - 234.1239871;,2987.293487298 + 84.7928748123 + 234 + 234.293847293 +
39847 - 29384798 - 234.1239871;,2987.293487298 + 84.7928748123 + 234 + 234.293847293 +
39847 - 29384798 - 234.1239871;,2987.293487298 + 84.7928748123 + 234 + 234.293847293 +
39847 - 29384798 - 234.1239871;,2987.293487298 + 84.7928748123 + 234 + 234.293847293 +
39847 - 29384798 - 23

ans =

    17    24     1     8    15
    23     5     7    14    16
     4     6    13    20    22
    10    12    19    21     3
    11    18    25     2     9

Elapsed time is 0.003924 seconds.
```

File: scripts/vim-matlab-server.py (hunk 1 above)
Comment (tdy) [inline scripts/vim-matlab-server.py]: Ah yes, thanks 👍

**Comment by daeyun** [comment]
Good catch, I think there was a race condition. I'm not seeing this anymore after e7ca8a44f. Let me know if you see other issues, thanks 👍 👍  looking forward to using this feature

**Pull request merged**